### ISSUE TITLE
Google security provider (ProviderInstaller) is flaky when calls are canceled

### ISSUE THREAD

**Issue by aliaksandr-sheiko89** [square/okhttp#4037]
Used combination OkHttp (3.10.0), RxJava (2.1.4) and Retrofit (2.3.0) libs. 

The issue reproducible on devices Samsung Galaxy A8(OS 7.1.1), Galaxy S8(OS 8.1), Galaxy S7(OS 7.0), Galaxy S5(OS 4.4.4).

Use case:
1) user opens screen with few tabs.  comment: here a bunch of requests sent
2) quickly switches to any other tab. comment: all requests needed for previous tab disposed (cancelled in terms of OhHttp)
3) screen contains few independent parts on UI (based on server's responses) and usually one of them loading too long

In logs I see that required requests are sent only after "java.net.SocketTimeoutException: timeout" printed and for our application it's 30 seconds.

Stacktrace below is for catch block okhttp3.RealCall.execute(okhttp3.RealCall:81)

 0 = {StackTraceElement@8913} "com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket$SSLInputStream.read(:com.google.android.gms@12685025@12.6.85 (040408-197041431):8)"
 1 = {StackTraceElement@8914} "okio.Okio$2.read(Okio.java:140)"
 2 = {StackTraceElement@8915} "okio.AsyncTimeout$2.read(AsyncTimeout.java:237)"
 3 = {StackTraceElement@8916} "okio.RealBufferedSource.indexOf(RealBufferedSource.java:355)"
 4 = {StackTraceElement@8917} "okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:227)"
 5 = {StackTraceElement@8918} "okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:215)"
 6 = {StackTraceElement@8919} "okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)"
 7 = {StackTraceElement@8920} "okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)"
 8 = {StackTraceElement@8921} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 9 = {StackTraceElement@8922} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)"
 10 = {StackTraceElement@8923} "ru.beeline.tve.android.utils.network.HttpClientBuilderFactory$BasicInterceptor.intercept(HttpClientBuilderFactory.java:64)"
 11 = {StackTraceElement@8924} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 12 = {StackTraceElement@8925} "okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)"
 13 = {StackTraceElement@8926} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 14 = {StackTraceElement@8927} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)"
 15 = {StackTraceElement@8928} "okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)"
 16 = {StackTraceElement@8929} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 17 = {StackTraceElement@8930} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)"
 18 = {StackTraceElement@8931} "okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)"
 19 = {StackTraceElement@8932} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 20 = {StackTraceElement@8933} "okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)"
 21 = {StackTraceElement@8934} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 22 = {StackTraceElement@8935} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)"
 23 = {StackTraceElement@8936} "ru.beeline.tve.android.utils.network.HttpClientBuilderFactory$ConnectivityInterceptor.intercept(HttpClientBuilderFactory.java:83)"
 24 = {StackTraceElement@8937} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 25 = {StackTraceElement@8938} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)"
 26 = {StackTraceElement@8939} "okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:213)"
 27 = {StackTraceElement@8940} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)"
 28 = {StackTraceElement@8941} "okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)"
 29 = {StackTraceElement@8942} "okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)"
 30 = {StackTraceElement@8943} "okhttp3.RealCall.execute(RealCall.java:77)"
 31 = {StackTraceElement@8944} "retrofit2.OkHttpCall.execute(OkHttpCall.java:180)"
 32 = {StackTraceElement@8945} "retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:41)"
 33 = {StackTraceElement@8946} "io.reactivex.Observable.subscribe(Observable.java:10903)"
 34 = {StackTraceElement@8947} "retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:34)"
 35 = {StackTraceElement@8948} "io.reactivex.Observable.subscribe(Observable.java:10903)"
 36 = {StackTraceElement@8949} "io.reactivex.internal.operators.observable.ObservableSingleSingle.subscribeActual(ObservableSingleSingle.java:35)"
 37 = {StackTraceElement@8950} "io.reactivex.Single.subscribe(Single.java:2707)"
 38 = {StackTraceElement@8951} "io.reactivex.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:36)"
 39 = {StackTraceElement@8952} "io.reactivex.Single.subscribe(Single.java:2707)"
 40 = {StackTraceElement@8953} "io.reactivex.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess(SingleFlatMap.java:84)"
 41 = {StackTraceElement@8954} "io.reactivex.internal.operators.single.SingleCreate$Emitter.onSuccess(SingleCreate.java:68)"
 42 = {StackTraceElement@8955} "ru.beeline.tve.android.domain.service.ApiService.lambda$getToken$16$ApiService(ApiService.java:596)"
 43 = {StackTraceElement@8956} "ru.beeline.tve.android.domain.service.ApiService$$Lambda$15.subscribe(Unknown Source)"
 44 = {StackTraceElement@8957} "io.reactivex.internal.operators.single.SingleCreate.subscribeActual(SingleCreate.java:39)"
 45 = {StackTraceElement@8958} "io.reactivex.Single.subscribe(Single.java:2707)"
 46 = {StackTraceElement@8959} "io.reactivex.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:36)"
 47 = {StackTraceElement@8960} "io.reactivex.Single.subscribe(Single.java:2707)"
 48 = {StackTraceElement@8961} "io.reactivex.internal.operators.single.SingleToFlowable.subscribeActual(SingleToFlowable.java:37)"
 49 = {StackTraceElement@8962} "io.reactivex.Flowable.subscribe(Flowable.java:12995)"
 50 = {StackTraceElement@8963} "io.reactivex.Flowable.subscribe(Flowable.java:12941)"
 51 = {StackTraceElement@8964} "io.reactivex.internal.operators.flowable.FlowableRepeatWhen$WhenReceiver.onNext(FlowableRepeatWhen.java:101)"
 52 = {StackTraceElement@8965} "io.reactivex.internal.operators.flowable.FlowableRetryWhen.subscribeActual(FlowableRetryWhen.java:62)"
 53 = {StackTraceElement@8966} "io.reactivex.Flowable.subscribe(Flowable.java:12995)"
 54 = {StackTraceElement@8967} "io.reactivex.internal.operators.flowable.FlowableSingleSingle.subscribeActual(FlowableSingleSingle.java:39)"
 55 = {StackTraceElement@8968} "io.reactivex.Single.subscribe(Single.java:2707)"
 56 = {StackTraceElement@8969} "io.reactivex.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess(SingleFlatMap.java:84)"
 57 = {StackTraceElement@8970} "io.reactivex.internal.operators.single.SingleCreate$Emitter.onSuccess(SingleCreate.java:68)"
 58 = {StackTraceElement@8971} "ru.beeline.tve.android.domain.service.ApiService.lambda$getKalturaSevice$13$ApiService(ApiService.java:507)"
 59 = {StackTraceElement@8972} "ru.beeline.tve.android.domain.service.ApiService$$Lambda$12.subscribe(Unknown Source)"
 60 = {StackTraceElement@8973} "io.reactivex.internal.operators.single.SingleCreate.subscribeActual(SingleCreate.java:39)"
 61 = {StackTraceElement@8974} "io.reactivex.Single.subscribe(Single.java:2707)"
 62 = {StackTraceElement@8975} "io.reactivex.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:36)"
 63 = {StackTraceElement@8976} "io.reactivex.Single.subscribe(Single.java:2707)"
 64 = {StackTraceElement@8977} "io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)"
 65 = {StackTraceElement@8978} "io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:452)"
 66 = {StackTraceElement@8979} "io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:61)"
 67 = {StackTraceElement@8980} "io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:52)"
 68 = {StackTraceElement@8981} "java.util.concurrent.FutureTask.run(FutureTask.java:237)"
 69 = {StackTraceElement@8982} "java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)"
 70 = {StackTraceElement@8983} "java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)"
 71 = {StackTraceElement@8984} "java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)"
 72 = {StackTraceElement@8985} "java.lang.Thread.run(Thread.java:762)"
 
Note: In our application we turn on ProviderInstaller.installIfNeeded to enable TLS1.2 on Android 4.4 (and following Google documentation improve connection security https://developer.android.com/training/articles/security-gms-provider) and after removing this action issue disappeared. First line in stack trace ("com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket$SSLInputStream.read(:com.google.android.gms@12685025@12.6.85 (040408-197041431):8)") is consequence of this usage.
 
 Here is a test project to reproduce it: https://github.com/aliaksandr-sheiko89/OkHttpIssue
 NOTE: sometimes it needs few attempt to reproduce.

**Comment by swankjesse** [comment]
To clarify: cancellation is slow or broken with the GMS provider?

**Comment by aliaksandr-sheiko89** [comment]
Cancellation of request sometimes makes following request (witch reuse from pool connection used by cancelled request) delayed for OkHttpClient.readTimeout period

**Comment by swankjesse** [comment]
FYI @kruton

**Comment by kruton** [comment]
You no longer had the issue when you removed the call to `ProviderInstaller.installIfNeeded` in all versions you reported reproducability (7.1.1, 8.1, 7.0, 4.4.4) or just 4.4?

@flooey

**Comment by aliaksandr-sheiko89** [comment]
In all versions

**Comment by yschimke** [comment]
@aliaksandr-sheiko89 we've had some fixes around this area, could you test with https://oss.jfrog.org/artifactory/libs-snapshot/com/squareup/okhttp3/okhttp/3.11.0-SNAPSHOT/

**Comment by aliaksandr-sheiko89** [comment]
sure, will try it today and share result after

**Comment by aliaksandr-sheiko89** [comment]
@yschimke reproduced for build 166 on Samsung A8 (os 7.1.1)

**Comment by yschimke** [comment]
I might look into this with android-test module.  Both testing with various socket providers and also more stress testing e.g. cancellation.

**Comment by yschimke** [comment]
Won't fix.  I think this is something that should be owned by Google Conscrypt, not us.
